src/nccl_ofi_net.c

```diff
@@ -188,34 +189,95 @@ int nccl_net_ofi_create_plugin(nccl_net_ofi_plugin_t **plugin_p)
 		NCCL_OFI_INFO(NCCL_INIT | NCCL_NET, "Using transport protocol %s (user set)",
 			      nccl_ofi_selected_protocol);
-	} else {
-		NCCL_OFI_INFO(NCCL_INIT | NCCL_NET, "Using transport protocol %s",
+	} else if (nccl_ofi_selected_protocol != NULL) {
+		NCCL_OFI_INFO(NCCL_INIT | NCCL_NET, "Using transport protocol %s (platform set)",
 			      nccl_ofi_selected_protocol);
 	}
 
-	if (0 == strcasecmp(nccl_ofi_selected_protocol, "SENDRECV")) {
-		ret = nccl_net_ofi_sendrecv_init(provider_filter, plugin_p);
-		if (ret != 0) {
-			NCCL_OFI_WARN("Failed to initialize sendrecv protocol");
+	if (nccl_ofi_selected_protocol != NULL) {
+		bool dummy;
+
+		if (0 == strcasecmp(nccl_ofi_selected_protocol, "SENDRECV")) {
+			ret = nccl_net_ofi_sendrecv_init(provider_filter, &plugin);
+			if (ret != 0) {
+				NCCL_OFI_WARN("Failed to initialize sendrecv protocol");
+				goto exit;
+			}
+		} else if (0 == strcasecmp(nccl_ofi_selected_protocol, "RDMA")) {
+			ret = nccl_net_ofi_rdma_init(provider_filter, &plugin, &dummy);
+			if (ret != 0) {
+				NCCL_OFI_WARN("Failed to initialize rdma protocol");
+				goto exit;
+			}
+		} else {
+			NCCL_OFI_WARN("Unable to find plugin protocol %s", nccl_ofi_selected_protocol);
+			ret = -ENOTSUP;
 			goto exit;
 		}
-	} else if (0 == strcasecmp(nccl_ofi_selected_protocol, "RDMA")) {
-		ret = nccl_net_ofi_rdma_init(provider_filter, plugin_p);
+	} else {
+		bool have_multiple_rails = false;
+		nccl_net_ofi_plugin_t *rdma_plugin = NULL, *sendrecv_plugin = NULL;
+
+		ret = nccl_net_ofi_rdma_init(provider_filter, &rdma_plugin, &have_multiple_rails);
 		if (ret != 0) {
-			NCCL_OFI_WARN("Failed to initialize rdma protocol");
+			NCCL_OFI_TRACE(NCCL_INIT | NCCL_NET,
+				       "Failed to initialize rdma protocol: %d", fi_strerror(-ret));
+			have_multiple_rails = false;
+			rdma_plugin = NULL;
+		}
+
+		if (!have_multiple_rails) {
+			ret = nccl_net_ofi_sendrecv_init(provider_filter, &sendrecv_plugin);
+			if (ret != 0) {
+				NCCL_OFI_TRACE(NCCL_INIT | NCCL_NET,
+					       "Failed to initialized rdma protocol: %d", fi_strerror(-ret));
+				sendrecv_plugin = NULL;
+			}
+		}
+
+		if (have_multiple_rails && rdma_plugin != NULL) {
+			nccl_ofi_selected_protocol = "RDMA";
+			plugin = rdma_plugin;
+			if (sendrecv_plugin != NULL) {
+				sendrecv_plugin->release_plugin(sendrecv_plugin);
+			}
+		} else {
+			nccl_ofi_selected_protocol = "SENDRECV";
+			plugin = sendrecv_plugin;
+			if (rdma_plugin != NULL) {
+				rdma_plugin->release_plugin(rdma_plugin);
+			}
+		}
+
+		if (nccl_ofi_selected_protocol == NULL) {
+			NCCL_OFI_WARN("Unable to find a protocol that worked.  Failing initialization.");
+			ret = -EINVAL;
 			goto exit;
 		}
-	} else {
-		NCCL_OFI_WARN("Unable to find plugin protocol %s", nccl_ofi_selected_protocol);
-		ret = -ENOTSUP;
+
+		NCCL_OFI_INFO(NCCL_INIT | NCCL_NET, "Using transport protocol %s",
+			      nccl_ofi_selected_protocol);
+	}
+
+	ret = plugin->complete_init(plugin);
```

Comment:
Nit: "complete_init" suggests that's the last thing we do for initialization, but we're calling get_ep() after this and creating network resources.   Marked nitpick, because I don't have a suggestion for how else to do it, since get_ep is proto agnostic and we'd have to duplicate that logic if it moves into complete_init.

Reply:
well, it is the last initialization step the plugin does.  The get_ep() calls aren't required for initialization of the plugin to work correctly.